dev(toolchain): Update toolchain and versions

big-r81 opened the pull request:
- Bump build tools versions
- Install dev dependencies into own tools directory (default: c:\tools)
- Save download artifacts in separate directory
- Don't mix Spidermonkey bins with vcpkg bins

nickva commented on bin/variables.ps1:
```diff
@@ -10,38 +10,44 @@
+$smBuildUri = "https://github.com/big-r81/couchdb-sm/releases/download/v${smBuildVersion}/${smBuild}.tar.xz"
 $smBuildFile = Split-Path $smBuildUri -Leaf
-$smInstallPath = "C:\relax\vcpkg\installed\x64-windows"
+$smInstallPath = "${vcpkgBase}"
```
good catch

nickva commented on bin/variables.ps1:
```diff
@@ -10,38 +10,44 @@
 # the License.
 
+# Installation directory for build tools
+$toolsDir = "C:\tools"
```
good idea, @big-r81

nickva commented on bin/install_dependencies.ps1:
```diff
@@ -13,18 +13,23 @@
+
 # Exclude c:\relax from MS defender to speed up things
 Add-MpPreference -ExclusionPath "C:\relax"
+Add-MpPreference -ExclusionPath "${toolsDir}"
```
very nice!

big-r81 commented:
@nickva, @pgj 
Hey, I updated this PR and added some handling of the downloaded artifacts. Maybe you can test this PR on a fresh Windows instalation and give feedback (if you have time).

Gábor: If this PR is merged, I would additional test your Clouseau PR!

pgj commented:
Thanks @big-r81 — yes, I have been planning to test your PR today or tomorrow.  I will get back to you on this once I have done it.

big-r81 commented:
Ok, the same error did I encountered too. It seems the path to lib dir of vcpkg isn’t available for the port compiler. I think we have to set it from an env variable and append maybe [here](https://github.com/apache/couchdb/blob/b85d4bb143550ad7227152c3f5e2fc6a6cf98f96/src/couch/rebar.config.script#L229).

I think there needs some clean of the compiler/linker switches for Windows.

pgj commented:
The libraries requested by the linker is on the `LIBPATH` (cf. `echo $env:LIBPATH`) but the full `LIBPATH` somehow does not make it to the site of invocation.

nickva commented:
@big-r81 unfortunately my windows instance is a physical machine and re-installing it probably won't work. I never did it just came with the laptop.

big-r81 commented:
> The libraries requested by the linker is on the `LIBPATH` (cf. `echo $env:LIBPATH`) but the full `LIBPATH` somehow does not make it to the site of invocation.

Mhh, we set explicitly the LIBPATH and yes it’s not catched up by the port compiler …

pgj commented:
Please remember to update the `README` file as well.  It has some version numbers included.

pgj commented:
> > The libraries requested by the linker is on the `LIBPATH` (cf. `echo $env:LIBPATH`) but the full `LIBPATH` somehow does not make it to the site of invocation.
> 
> Mhh, we set explicitly the LIBPATH and yes it’s not catched up by the port compiler …

I have just learnt that the linker uses the `LIB` environment variable instead.

big-r81 commented:
> I have just learnt that the linker uses the `LIB` environment variable instead.

`LIB` and `LIBPATH` should be the same and I’m pondering why it doesn’t work now…

pgj commented:
I think I have found something curious.  One of the directories on the library path (in the `LIB` environment variable) has _some_ of the ICU libraries, which — I believe — confuses the linker.  This is the "Windows Kits" library directory: `C:\Program Files (x86)\Windows Kits\10\Lib\10.0.2261.0\um\x64` which has `icu.lib`, `icuin.lib`, and `icuuc.lib`.

pgj commented:
https://learn.microsoft.com/en-us/dotnet/core/extensions/globalization-icu#icu-on-windows

big-r81 commented:
> I think I have found something curious. One of the directories on the library path (in the `LIB` environment variable) has _some_ of the ICU libraries, which — I believe — confuses the linker. This is the "Windows Kits" library directory: `C:\Program Files (x86)\Windows Kits\10\Lib\10.0.2261.0\um\x64` which has `icu.lib`, `icuin.lib`, and `icuuc.lib`.

1. A interesting. I have an idea. I changed the logic, how the env vars `LIB` and `LIBPATH` are created. Now the paths are appended to the end of teh env vars and before they were prepended. I will change the order and test again.

2. Maybe we don't need the extra vcpkg-icu libs anymore?

pgj commented:
Right, we do not necessarily need the `icu` package via `vcpkg`.  However, according to the documentation I linked above in my comment, the built-in ICU has to be used somewhat differently.  There is a single unified `icu.h` and `icu.lib` and that have to be added to the corresponding places.  See https://github.com/microsoft/icu for more information.

With the `icu` package having removed, and having fixed the header and library name references, I was able to build the code on Windows again.  Note that Windows ships only ICU 72 but I guess that is enough for us.

big-r81 commented:
So, I updated the PR and changed the order of the appended (prepended) paths. It should work now!

pgj commented:
One minor note.  Maybe that is unrelated but it bugs me a bit.

After installing `git` via `choco`, the path for the binaries is not (automatically) added to the `PATH`:

```console
mkdir C:\relax\
cd C:\relax\
Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))
choco feature enable -n allowGlobalConfirmation
choco install git
```

Thus the invocations of `git` does not work out-of-the-box as the subsequent instructions would suggest:

```console
git config --global core.autocrlf false
git clone https://github.com/apache/couchdb-glazier
```

To make them work, I had insert the following line:

```console
$env:PATH = $env:PATH + ';C:\Program Files\Git\bin'
```

pgj commented:
I see these error messages at the end:

```console
The following packages will be built and installed:
    icu[core,tools]:x64-windows -> 73.1
error: in triplet x64-windows: Unable to find a valid Visual Studio instance
Could not locate a complete Visual Studio instance
The following paths were examined for Visual Studio instances:
  C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary/Build\vcvarsall.bat

Installing OpenSSL ...
Computing installation plan...
The following packages will be built and installed:
    openssl:x64-windows -> 3.1.3
  * vcpkg-cmake:x64-windows -> 2023-05-04
  * vcpkg-cmake-config:x64-windows -> 2022-02-06#1
  * vcpkg-cmake-get-vars:x64-windows -> 2023-03-02
Additional packages (*) will be modified to complete this operation.
error: in triplet x64-windows: Unable to find a valid Visual Studio instance
Could not locate a complete Visual Studio instance
The following paths were examined for Visual Studio instances:
  C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary/Build\vcvarsall.bat

Downloading SpiderMonkey ...
Installing SpiderMonkey ...
VALUE: &archiver.TarXz{Tar:(*archiver.Tar)(0xc00005c720)}
copy : Container cannot be copied onto existing leaf item.
At C:\relax\couchdb-glazier\bin\install_dependencies.ps1:83 char:1
+ copy -Recurse -Force "$smBuild\*" $vcpkgBase
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (C:\relax\Windows-mozjs-91\include:String) [Copy-Item], PSArgumentException
    + FullyQualifiedErrorId : CopyContainerItemToLeafError,Microsoft.PowerShell.Commands.CopyItemCommand

copy : Container cannot be copied onto existing leaf item.
At C:\relax\couchdb-glazier\bin\install_dependencies.ps1:83 char:1
+ copy -Recurse -Force "$smBuild\*" $vcpkgBase
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (C:\relax\Windows-mozjs-91\lib:String) [Copy-Item], PSArgumentException
    + FullyQualifiedErrorId : CopyContainerItemToLeafError,Microsoft.PowerShell.Commands.CopyItemCommand
```

In result, SpiderMonkey cannot be found on configure:

```console
.\configure.ps1 -WithClouseau
ERROR: SpiderMonkey 91 is not found. Please specify with -SpiderMonkeyVersion.
```

big-r81 commented:
> I see these error messages at the end:
> 
> ```
> The following packages will be built and installed:
>     icu[core,tools]:x64-windows -> 73.1
> error: in triplet x64-windows: Unable to find a valid Visual Studio instance
> Could not locate a complete Visual Studio instance
> The following paths were examined for Visual Studio instances:
>   C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary/Build\vcvarsall.bat
> 
> Installing OpenSSL ...
> Computing installation plan...
> The following packages will be built and installed:
>     openssl:x64-windows -> 3.1.3
>   * vcpkg-cmake:x64-windows -> 2023-05-04
>   * vcpkg-cmake-config:x64-windows -> 2022-02-06#1
>   * vcpkg-cmake-get-vars:x64-windows -> 2023-03-02
> Additional packages (*) will be modified to complete this operation.
> error: in triplet x64-windows: Unable to find a valid Visual Studio instance
> Could not locate a complete Visual Studio instance
> The following paths were examined for Visual Studio instances:
>   C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary/Build\vcvarsall.bat
> 
> Downloading SpiderMonkey ...
> Installing SpiderMonkey ...
> VALUE: &archiver.TarXz{Tar:(*archiver.Tar)(0xc00005c720)}
> copy : Container cannot be copied onto existing leaf item.
> At C:\relax\couchdb-glazier\bin\install_dependencies.ps1:83 char:1
> + copy -Recurse -Force "$smBuild\*" $vcpkgBase
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidArgument: (C:\relax\Windows-mozjs-91\include:String) [Copy-Item], PSArgumentException
>     + FullyQualifiedErrorId : CopyContainerItemToLeafError,Microsoft.PowerShell.Commands.CopyItemCommand
> 
> copy : Container cannot be copied onto existing leaf item.
> At C:\relax\couchdb-glazier\bin\install_dependencies.ps1:83 char:1
> + copy -Recurse -Force "$smBuild\*" $vcpkgBase
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidArgument: (C:\relax\Windows-mozjs-91\lib:String) [Copy-Item], PSArgumentException
>     + FullyQualifiedErrorId : CopyContainerItemToLeafError,Microsoft.PowerShell.Commands.CopyItemCommand
> ```
> 
> In result, SpiderMonkey cannot be found on configure:
> 
> ```
> .\configure.ps1
> ERROR: SpiderMonkey 91 is not found. Please specify with -SpiderMonkeyVersion.
> ```

It seems, your branch is updated. The copy commands doesn't exists anymore ...

pgj commented:
Aw, you are right.  I was using the wrong branch :-(

big-r81 commented:
> One minor note. Maybe that is unrelated but it bugs me a bit.
> 
> After installing `git` via `choco`, the path for the binaries is not (automatically) added to the `PATH`:
> 
> ```
> mkdir C:\relax\
> cd C:\relax\
> Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))
> choco feature enable -n allowGlobalConfirmation
> choco install git
> ```
> 
> Thus the invocations of `git` does not work out-of-the-box as the subsequent instructions would suggest:
> 
> ```
> git config --global core.autocrlf false
> git clone https://github.com/apache/couchdb-glazier
> ```
> 
> To make them work, I had insert the following line:
> 
> ```
> $env:PATH = $env:PATH + ';C:\Program Files\Git\bin'
> ```

Oh good point, we need to call `refreshenv`, but it isn't available at this time, so we would need to use an explicit workaround like (here)[https://github.com/apache/couchdb-glazier/blob/e13f3eb48f4562da2e62d3486a0b6a60cde2d47b/bin/install_dependencies.ps1#L32-L37]. The benefit would be, that we don't need to hardcode the path.

We should add a `bottstrap` script, which does all these steps automatically...

big-r81 commented:
Hey Gábor, could you test this "bootstrap" script, instead of doing the "first" steps manually?

```powershell
$relaxDir = "C:\relax"

if (-not (Test-Path -Path $relaxDir)) {
  New-Item -Path $relaxDir -ItemType Directory
}

Set-Location $relaxDir

Set-ExecutionPolicy Bypass -Scope Process -Force; [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; iex ((New-Object System.Net.WebClient).DownloadString('https://community.chocolatey.org/install.ps1'))
choco feature enable -n allowGlobalConfirmation
choco install git

# Explicit workaround that refresnenv is working for this powershell shell
$ChocolateyProfile = "$env:ChocolateyInstall\helpers\chocolateyProfile.psm1"
if (Test-Path($ChocolateyProfile)) {
  Import-Module "$ChocolateyProfile"
}
refreshenv

git config --global core.autocrlf false
git clone https://github.com/apache/couchdb-glazier
&.\couchdb-glazier\bin\install_dependencies.ps1
```

pgj commented:
Do not we want it make the bootstrap script parametric?  One may want to use it against different branches and forks of `apache/couchdb-glazier`.

big-r81 commented:
> Do not we want it make the bootstrap script parametric? One may want to use it against different branches and forks of `apache/couchdb-glazier`.

Mhhh, my idea was that a user just run it (with a doubelick or per shell) and all is setup, like the bootstrap script of vcpkg ...

pgj commented:
> Mhhh, my idea was that a user just run it (with a doubelick or per shell) and all is setup, like the bootstrap script of vcpkg ...

It can come handy for developers as well.

big-r81 commented:
> It can come handy for developers as well. You can always default to `https://github.com/apache/couchdb-glaizer`.

I think it's a good idea. The question is now, if we only add the simple bootstrap script for the moment and do the additional features in a separate PR? The initial idea was to update and simplify the build steps for CouchDB in Windows.

pgj commented:
I do not insist on addressing all these issues in a single PR.  That said, let us make the bootstrap script a separate commit and PR.

pgj reviewed (approved):
Great work!

big-r81 commented:
> I do not insist on addressing all these issues in a single PR. That said, let us make the bootstrap script a separate commit and PR.

Yes, sounds good. So for the current state, all changes works for you on a fresh install?

pgj commented:
Yes, they do.
